proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+                                tcbPtr \<noteq> ksIdleThread s \<and> from \<noteq> ksIdleThread s"
+                    in hoare_strengthen_post)
+        apply wpsimp+
+  apply (auto simp: obj_at'_def projectKO_eq projectKO_sc valid_idle'_def)
+  done
+
+lemma schedContextDonate_valid_release_queue'[wp]:
+  "schedContextDonate scPtr tcbPtr \<lbrace>valid_release_queue'\<rbrace>"
+  unfolding schedContextDonate_def
+  by (wpsimp wp: threadSet_valid_release_queue' simp: obj_at_ko_at'_eq valid_release_queue'_def
+      | rule hoare_strengthen_post[where Q="\<lambda>_. valid_release_queue'"])+
+
+lemma schedContextDonate_assume_asserts:
+  "\<lbrace>\<lambda>s. P s \<and> sc_at' scPtr s \<and> sym_refs (state_refs_of' s)\<rbrace> schedContextDonate scPtr tcbPtr \<lbrace>Q\<rbrace>
+    \<Longrightarrow> \<lbrace>P\<rbrace> schedContextDonate scPtr tcbPtr \<lbrace>Q\<rbrace>"
+  apply (simp only: schedContextDonate_def)
+  apply (rule hoare_strengthen_pre_via_assert_backward[rotated])
+   apply (simp only: pred_conj_def)
+  apply (rule hoare_seq_ext[OF _ stateAssert_sp])
+  apply (rule hoare_K_bind)
+  apply (rule hoare_seq_ext[OF _ get_sc_sp'])
+  apply (rule_tac hoare_weaken_pre[OF hoare_pre_cont])
+  apply (clarsimp simp: obj_at'_def sym_refs_asrt_def)
+  done
+
+lemma schedContextDonate_invs':
+  "\<lbrace>\<lambda>s. invs' s \<and> bound_sc_tcb_at' ((=) None) tcbPtr s \<and>
+        ex_nonz_cap_to' scPtr s \<and> ex_nonz_cap_to' tcbPtr s\<rbrace>
+   schedContextDonate scPtr tcbPtr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (simp only: invs'_def valid_state'_def)
+  apply (rule schedContextDonate_assume_asserts)
+  apply (wpsimp wp: schedContextDonate_valid_pspace' schedContextDonate_valid_release_queue
+                    schedContextDonate_valid_queues schedContextDonate_valid_queues'
+                    schedContextDonate_valid_idle' schedContextDonate_if_live_then_nonz_cap')
+  apply (clarsimp simp: obj_at'_def projectKO_eq projectKO_sc)
+  apply (drule_tac ko=obj in sym_refs_ko_atD'[rotated, where p=scPtr])
+   apply (auto dest!: global'_sc_no_ex_cap
+               simp: ko_wp_at'_def obj_at'_def projectKO_eq projectKO_tcb
+                     pred_tcb_at'_def valid_idle'_def idle_tcb'_def refs_of_rev')
+  done
+
+lemma valid_tcb_state'_update[simp]:
+  "valid_tcb_state' ts (s\<lparr>ksReadyQueues := qs\<rparr>) = valid_tcb_state' ts s"
+  "valid_tcb_state' ts (s\<lparr>ksReadyQueuesL1Bitmap := l1b\<rparr>) = valid_tcb_state' ts s"
+  "valid_tcb_state' ts (s\<lparr>ksReadyQueuesL2Bitmap := l2b\<rparr>) = valid_tcb_state' ts s"
+  "valid_tcb_state' ts (s\<lparr>ksSchedulerAction := sa\<rparr>) = valid_tcb_state' ts s"
+  by (auto simp: valid_tcb_state'_def valid_bound_obj'_def split: thread_state.splits option.splits)
+
+crunches addToBitmap, setQueue
+  for valid_tcb_state'[wp]: "valid_tcb_state' ts"
+
+lemma setSchedContext_obj_at':
```

Comment:
Are you able to move this and the following lemmas to the `simple_ko'` locale? If you can it would let you avoid duplicating it for the various different types and constants.

Reply:
This looks like a missing variant of `simple_ko'.obj_at'_strongest`. There are "nice" "simple" variants for `P (ko_at' ko p s)` but not for `P (obj_at' Q p s)`, so it makes sense to add a `simple_ko'.obj_at'` which has the form of your `setSchedContext_obj_at'`: this is probably the form people find most useful/familiar.